src/Materials/Textures/Packer/packer.ts

```diff
@@ -0,0 +1,551 @@
+                            let t: Texture = (mat as any)[index];
+                            if (t !== null) {
+
+                                if (!(this.sets as any)[this.options.map[j]]) {
+                                    (this.sets as any)[this.options.map[j]] = true;
+                                }
+
+                                if (this.options.disposeSources) {
+                                    this._disposeList.push(t);
+                                }
+                            }
+                        }
+
+                        if (done == this.meshes.length) {
+                            this._createFrames(resolve);
+                        }
+                    }
+                };
+
+                for (let i = 0; i < this.meshes.length; i++) {
+                    let mesh = this.meshes[i];
+                    let material: Nullable<Material> | Material = mesh.material;
+                    if (!material) {
+                        return new Error('Mesh has no Material Assigned!');
+                    }
+
+                    material.forceCompilationAsync(mesh).then(() => {
+                        doneCheck((material as Material));
+                    });
+
+                }
+            }catch (e) {
+                return reject(e);
+            }
+        });
+
+    return;
+    }
+
+    /**
+    * Starts the package process
+    * @param resolve The promises resolution function
+    * @returns TexturePacker
+    */
+    private _createFrames(resolve: Function) {
+
+        let dtSize = this._calculateSize();
+        let dtUnits = (new Vector2(1, 1)).divide(dtSize);
+        let doneCount = 0;
+        let expecting = this._disposeList.length;
+        let meshLength = this.meshes.length;
+
+        let sKeys = Object.keys(this.sets);
+        for (let i = 0; i < sKeys.length; i++) {
+            let setName = sKeys[i];
+
+            let dt = new DynamicTexture(this.name + '.TexturePack.' + setName + 'Set',
+                    {width: dtSize.x, height: dtSize.y},
+                    this.scene,
+                    true, //Generate Mips
+                    Texture.TRILINEAR_SAMPLINGMODE,
+                    Engine.TEXTUREFORMAT_RGBA
+                );
+
+            let dtx = dt.getContext();
+            dtx.fillStyle = 'rgba(0,0,0,0)';
+            dtx.fillRect(0, 0, dtSize.x, dtSize.y) ;
+            dt.update();
+            (this.sets as any)[setName] = dt;
+        }
+
+        let baseSize = this.options.frameSize || 256;
+        let padding = this._paddingValue;
+        let tcs = baseSize + (2 * padding);
+
+        const done = () => {
+            console.log("Compilation Done");
+            this._calculateMeshUVFrames(baseSize, padding, dtSize, dtUnits, this.options.updateInputMeshes || false);
+        };
+
+        //Update the Textures
+        for (let i = 0; i < meshLength; i++) {
+            let m = this.meshes[i];
+            let mat = m.material;

```

Comment:
What if they share the same material, are you not doing it multiple times?

Reply:
No, because the texture then already exists on the sets and so I just make sure the mesh gets set to the correct frames for that material.